### ISSUE TITLE
HTTP Issues

### ISSUE THREAD

**Issue by methodds** [GeneralMills/pytrends#47]
Hi,

trying to connect raises an error for me:

```
----> 8 connector = pyGTrends(google_username, google_password)
      9 
     10 # make request

/home/cs/anaconda3/lib/python3.5/site-packages/pytrends/pyGTrends.py in __init__(self, username, password)
     49         self.url_CookieCheck = 'https://www.google.com/accounts/CheckCookie?chtml=LoginDoneHtml'
     50         self.url_PrefCookie = 'http://www.google.com'
---> 51         self._connect()
     52 
     53     def _connect(self):

/home/cs/anaconda3/lib/python3.5/site-packages/pytrends/pyGTrends.py in _connect(self)
     73         params = urlencode(self.login_params).encode('utf-8')
     74         self.opener.open(self.url_ServiceLoginBoxAuth, params)
---> 75         self.opener.open(self.url_CookieCheck)
     76         self.opener.open(self.url_PrefCookie)
     77 

/home/cs/anaconda3/lib/python3.5/urllib/request.py in open(self, fullurl, data, timeout)
    469         for processor in self.process_response.get(protocol, []):
    470             meth = getattr(processor, meth_name)
--> 471             response = meth(req, response)
    472 
    473         return response

/home/cs/anaconda3/lib/python3.5/urllib/request.py in http_response(self, request, response)
    579         if not (200 <= code < 300):
    580             response = self.parent.error(
--> 581                 'http', request, response, code, msg, hdrs)
    582 
    583         return response

/home/cs/anaconda3/lib/python3.5/urllib/request.py in error(self, proto, *args)
    501             http_err = 0
    502         args = (dict, proto, meth_name) + args
--> 503         result = self._call_chain(*args)
    504         if result:
    505             return result

/home/cs/anaconda3/lib/python3.5/urllib/request.py in _call_chain(self, chain, kind, meth_name, *args)
    441         for handler in handlers:
    442             func = getattr(handler, meth_name)
--> 443             result = func(*args)
    444             if result is not None:
    445                 return result

/home/cs/anaconda3/lib/python3.5/urllib/request.py in http_error_302(self, req, fp, code, msg, headers)
    684         fp.close()
    685 
--> 686         return self.parent.open(new, timeout=req.timeout)
    687 
    688     http_error_301 = http_error_303 = http_error_307 = http_error_302

/home/cs/anaconda3/lib/python3.5/urllib/request.py in open(self, fullurl, data, timeout)
    469         for processor in self.process_response.get(protocol, []):
    470             meth = getattr(processor, meth_name)
--> 471             response = meth(req, response)
    472 
    473         return response

/home/cs/anaconda3/lib/python3.5/urllib/request.py in http_response(self, request, response)
    579         if not (200 <= code < 300):
    580             response = self.parent.error(
--> 581                 'http', request, response, code, msg, hdrs)
    582 
    583         return response

/home/cs/anaconda3/lib/python3.5/urllib/request.py in error(self, proto, *args)
    507         if http_err:
    508             args = (dict, 'default', 'http_error_default') + orig_args
--> 509             return self._call_chain(*args)
    510 
    511 # XXX probably also want an abstract factory that knows when it makes

/home/cs/anaconda3/lib/python3.5/urllib/request.py in _call_chain(self, chain, kind, meth_name, *args)
    441         for handler in handlers:
    442             func = getattr(handler, meth_name)
--> 443             result = func(*args)
    444             if result is not None:
    445                 return result

/home/cs/anaconda3/lib/python3.5/urllib/request.py in http_error_default(self, req, fp, code, msg, hdrs)
    587 class HTTPDefaultErrorHandler(BaseHandler):
    588     def http_error_default(self, req, fp, code, msg, hdrs):
--> 589         raise HTTPError(req.full_url, code, msg, hdrs, fp)
    590 
    591 class HTTPRedirectHandler(BaseHandler):

HTTPError: HTTP Error 400: Bad Request
```
Do you have an idea why the request does not work properly? I double checked my authentification credentials and these are correct.

Thanks and Cheers
Carsten

**Comment by shobute** [comment]
I am also seeing the same issue with Python 2. It worked on the 19th but not since then, with unchanged username/password and code.

**Comment by oshoma** [comment]
I get the same error with Python 3.4.3 i.e. the  call to `self.opener.open(self.url_CookieCheck)` fails. My authentication credentials haven't changed since last use. Perhaps Google has changed the behavior of that page.

**Comment by mcapogreco** [comment]
I am also seeing the same issue - checked login details and they are correct, plus checked latest version..
The google trends site still works as expected when using the url's and same account. Hope this can be resolved soon - such a useful tool!

**Comment by ericross2** [comment]
First off: I greatly appreciate the work done on pytrends. Thank you for providing this.

Pytrends worked perfectly for several weeks with zero glitches on Python 3.5.2. On Monday I received the same 400 error, and since then I have been stuck. I don't know what (if anything) changed. I too tried what @adstastic did in January (closed 400 error topic) and saw the exact same results (same csv file dump s/he saw). I also tried the pip install pytrends --"upgrade" you had recommended in March for Python 3.5.1 to ravimevcha from "Script stopped working.. 400 Bad Request error #30" but still the same 400 error.

I believe I've traced the issue to the self.opener.open(self.url_CookieCheck). My Google credentials are correct and work everywhere else (and I have no messages from Google relating to security or proving I am not a bot, and thus I don't believe Google locked my account. ) so I am at a loss as to how to proceed.

If I manually paste the google trends request url directly into the browser (I am using Opera; already logged into the same Google account for email/etc.) it gives me a good csv. I cleared the Opera cache, and don't show any cookies listed.

To be clear: I am only downloading between 20-50 trends TOTAL per day (each a list of up to five terms to be compared as a single trend/url/csv); this is not a mass-trend downloader for me.

I apologize if accidentally broke protocol but I had originally commented on this via a closed issue: HTTP Error 400: Bad Request #24. Actually there have been a number of similar comments on this closed issue in the past several days.

Any help is appreciated.

Thank you.

**Comment by lmclupr** [comment]
I am getting the same error.  The problem is that Google has changed its backend again.  This means that this module, without some serious changes no longer works.  From my initial investigation, it goes beyond simple URL change.

**Comment by dreyco676** [comment]
I will be be taking a look at this today.

This package is something I support in my freetime so I appreciate
understanding in the time it takes to fix.

**Comment by ericross2** [comment]
John,

I believe I speak for all on this forum that we are all greatly appreciative of your time and effort, and of your graciousness in offering it for use.

Thank you.

Eric

**Comment by lmclupr** [comment]
John, I find it extraordinary that guys like you maintain free packages
like pygtrend.  I am very grateful. Thanks

**Comment by dreyco676** [comment]
Working on it now. Looks like its finally time to move to Requests. I'm able to log into my Google Account now I just need to pass the cookie to the other report request.


@ericross2 thanks for your help. The reason it is working when you paste the URL is because you are most likely already logged into Google through your browser.

**Comment by dreyco676** [comment]
Got it. It's messy so I'll refactor after some coffee tomorrow but here is the gist in case I get hit by a meteor while I sleep.

class SessionGoogle:
    def __init__(self, url_login, url_auth, login, pwd):
        self.ses = requests.session()
        login_html = self.ses.get(url_login)
        soup_login = BeautifulSoup(login_html.content, "lxml").find('form').find_all('input')
        dico = {}
        for u in soup_login:
            if u.has_attr('value'):
                dico[u['name']] = u['value']
        # override the inputs with out login and pwd:
        dico['Email'] = login
        dico['Passwd'] = pwd
        self.ses.post(url_auth, data=dico)

    def get(self, URL):
        return self.ses.get(URL).text

url_login = "https://accounts.google.com/ServiceLogin"
url_auth = "https://accounts.google.com/ServiceLoginAuth"
session = SessionGoogle(url_login, url_auth, google_username, google_password)
req = session.get("http://www.google.com/trends/trendsReport?q=Pizza&hl=en-US&cmpt=q&content=1&export=1")
print(req)

**Comment by iethan** [comment]
I tried working on this and Google is logging me out. I adapted some of your code: 

```
from bs4 import BeautifulSoup

class SessionGoogle:
	def __init__(self, url_login, url_auth, login, pwd):
		self.url_login = url_login
		self.url_auth = url_auth
		self.login = login
		self.pwd = pwd
		self.ses = requests.session()
		self.login_html = self.ses.get(url_login)
		self.soup_login = BeautifulSoup(self.login_html.content, "lxml").find("form").find_all("input")
		self.dico = {}


	def log(self):

		for u in self.soup_login:
			if u.has_attr('value'):
				# print(u.encode('utf-8'))
				self.dico[u['name']] = u['value']
				# override the inputs with out login and pwd:
				self.dico['Email'] = self.login
				self.dico['Passwd'] = self.pwd
				self.ses.post(self.url_auth, data=self.dico)

	def get(self, URL):		
		return self.ses.get(URL).text

url_login = "https://accounts.google.com/ServiceLogin"
url_auth = "https://accounts.google.com/ServiceLoginAuth"
session = SessionGoogle(url_login, url_auth, username, password)
session.log()
req = session.get("http://www.google.com/trends/trendsReport?q=Pizza&hl=en-US&cmpt=q&content=1&export=1")
print(req)
```
It is automatically logging me out. Here is the output

```
<html><head><title>Redirecting</title>

<script type="text/javascript" language="javascript">
      // Accessing window.external members can cause IE to throw exceptions.
      // Any code that acesses window.external members must be try/catch wrapped
      /** @preserveTry */
      try {
        if (top == self) {
          if (window.gtbExternal) {
            window.gtbExternal.setM();
          } else {
            window.external.setM();
          }
        }
      }
      catch(err) {
      }
    </script>

<meta http-equiv="refresh" content="0; url=&#39;http://www.google.com/trends#cmpt=q&amp;q=Pizza&amp;hl=en-US&amp;content=1&#39;"></head>
<body bgcolor="#ffffff" text="#000000" link="#0000cc" vlink="#551a8b" alink="#ff0000"><script type="text/javascript" language="javascript">
    location.replace("http://www.google.com/trends#cmpt\x3dq\x26q\x3dPizza\x26hl\x3den-US\x26content\x3d1")
  </script></body></html>

```

The code should work so I'm not sure why it's logging me out. Any help would be greatly appreciated. And thank you again @dreyco676 for your help!

**Comment by dreyco676** [comment]
New version released 1.2.0 both on github and Pypi. _Functionality remains the same._ Next major release will use dictionaries for parameters passed in, I will keep both up on Pypi once that happens.

**Known issues:**
* Google will nag you that you've logged in from a new device each time.
